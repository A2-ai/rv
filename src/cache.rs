use std::fmt;
use std::fmt::Formatter;
use std::path::PathBuf;

#[derive(Debug, PartialEq, Clone)]
pub enum CacheEntry {
    Existing(PathBuf),
    NotFound(PathBuf),
    Expired(PathBuf),
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum InstallationStatus {
    Source,
    Binary,
    Both,
    Absent,
}

impl InstallationStatus {
    pub fn available(&self) -> bool {
        *self != InstallationStatus::Absent
    }

    pub fn binary_available(&self) -> bool {
        matches!(self, InstallationStatus::Binary | InstallationStatus::Both)
    }

    pub fn source_available(&self) -> bool {
        matches!(self, InstallationStatus::Source | InstallationStatus::Both)
    }
}

impl fmt::Display for InstallationStatus {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        let v = match self {
            InstallationStatus::Source => "source",
            InstallationStatus::Binary => "binary",
            InstallationStatus::Both => "source and binary",
            InstallationStatus::Absent => "absent",
        };
        write!(f, "{v}")
    }
}

pub trait Cache {
    /// This will either load the database for that repository or return None if we couldn't find
    /// it or it was expired.
    fn get_package_db_entry(&self, repo_url: &str) -> CacheEntry;

    /// Gets the status of a package coming from a package repository in the cache
    fn get_package_installation_status(
        &self,
        repo_url: &str,
        name: &str,
        version: &str,
    ) -> InstallationStatus;

    fn get_git_installation_status(&self, repo_url: &str, sha: &str) -> InstallationStatus;

    /// Gets the path to where a git repository should be cloned
    fn get_git_clone_path(&self, repo_url: &str) -> PathBuf;

    /// Gets the path to where a tarball package should be downloaded
    fn get_url_download_path(&self, url: &str) -> PathBuf;
}
