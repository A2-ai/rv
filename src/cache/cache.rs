use std::path::PathBuf;

#[derive(Debug, Clone)]
pub struct Cache {
    /// The cache directory.
    /// In practice it will be the OS own cache specific directory + `rv`
    root: PathBuf,
}

impl Cache {
    pub fn new() -> Self {

    }
}