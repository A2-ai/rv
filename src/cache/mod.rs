pub mod disk;
pub mod utils;

pub use disk::{DiskCache, InstallationStatus, PackagePaths};
