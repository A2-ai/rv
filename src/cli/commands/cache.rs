use std::path::Path;

use anyhow::Result;

use crate::{cli::{CliContext, RCommandLookup}, CacheInfo};

fn list_dirs(config_file: impl AsRef<Path>, log_enabled: bool) -> Result<CacheInfo> {
    let mut context = CliContext::new(config_file.as_ref(), RCommandLookup::Skip)?;
    if !log_enabled {
        context.show_progress_bar();
    }
    context.load_databases()?;

    Ok(())
}