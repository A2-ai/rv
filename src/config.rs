use std::collections::HashMap;
use std::path::Path;

use serde::Deserialize;

use crate::r_cmd::RCmd;
use crate::version::Version;

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(deny_unknown_fields)]
struct Author {
    name: String,
    email: String,
    #[serde(default)]
    maintainer: bool,
}

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Repository {
    pub alias: String,
    pub url: String,
}

#[derive(Debug, PartialEq, Clone, Deserialize)]
#[serde(untagged)]
pub enum DependencyKind {
    Simple(String),
    Detailed {
        name: String,
        repository: Option<String>,
        url: Option<String>,
        #[serde(default)]
        install_suggestions: bool,
        #[serde(default)]
        force_source: bool,
    },
}

impl DependencyKind {
    pub fn name(&self) -> &str {
        match self {
            DependencyKind::Simple(s) => s,
            DependencyKind::Detailed { name, .. } => name,
        }
    }

    pub fn repository(&self) -> Option<&str> {
        match self {
            DependencyKind::Simple(_) => None,
            DependencyKind::Detailed { repository, .. } => repository.as_deref(),
        }
    }

    pub fn force_source(&self) -> bool {
        match self {
            DependencyKind::Simple(_) => false,
            DependencyKind::Detailed { force_source, .. } => *force_source,
        }
    }

    pub fn install_suggestions(&self) -> bool {
        match self {
            DependencyKind::Simple(_) => false,
            DependencyKind::Detailed {
                install_suggestions,
                ..
            } => *install_suggestions,
        }
    }
}

#[derive(Debug, PartialEq, Clone, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub(crate) struct Project {
    name: String,
    #[serde(default)]
    description: String,
    version: Option<Version>,
    license: Option<String>,
    #[serde(default)]
    authors: Vec<Author>,
    #[serde(default)]
    keywords: Vec<String>,
    repositories: Vec<Repository>,
    #[serde(default)]
    suggests: Vec<DependencyKind>,
    r_version: Option<Version>,
    #[serde(default)]
    urls: HashMap<String, String>,
    #[serde(default)]
    dependencies: Vec<DependencyKind>,
}

#[derive(Debug, PartialEq, Clone, Deserialize)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
pub struct Config {
    pub(crate) project: Project,
}

impl Config {
    // TODO: handle errors later
    pub fn from_file<P: AsRef<Path>>(path: P) -> Self {
        // TODO: use a custom read file method that reports the filepath if it fails
        let content = std::fs::read_to_string(path).expect("TODO: handle error");
        toml::from_str(content.as_str()).expect("TODO: handle error")
    }

    /// Gets the R version we want to use in this project.
    /// This will default to whatever is set in the config if set, otherwise pick the output
    /// from the trait call
    pub fn get_r_version(&self, r_cli: impl RCmd) -> Version {
        if let Some(v) = &self.project.r_version {
            v.clone()
        } else {
            r_cli.version()
        }
    }

    pub fn repositories(&self) -> &[Repository] {
        &self.project.repositories
    }

    pub fn dependencies(&self) -> &[DependencyKind] {
        &self.project.dependencies
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn can_parse_valid_config_files() {
        let paths = std::fs::read_dir("src/tests/valid_config/").unwrap();
        for path in paths {
            // TODO: later it will return a res.
            let _ = Config::from_file(path.unwrap().path());
        }
    }
}
