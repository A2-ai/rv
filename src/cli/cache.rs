use std::path::PathBuf;
use std::time::SystemTime;

use anyhow::{anyhow, Context, Result};
use base64::{engine::general_purpose::STANDARD_NO_PAD, Engine as _};
use etcetera::BaseStrategy;
use fs_err as fs;

use crate::cache::InstallationStatus;
use crate::cli::utils::get_os_path;
use crate::system_info::SystemInfo;
use crate::version::Version;
use crate::{Cache, CacheEntry};

/// How long are the package databases cached for
/// Same default value as PKGCACHE_TIMEOUT:
/// https://github.com/r-lib/pkgcache?tab=readme-ov-file#package-environment-variables
const PACKAGE_TIMEOUT: u64 = 60 * 60;
const PACKAGE_TIMEOUT_ENV_VAR_NAME: &str = "PKGCACHE_TIMEOUT";
const PACKAGE_DB_FILENAME: &str = "packages.bin";

fn get_user_cache_dir() -> Option<PathBuf> {
    etcetera::base_strategy::choose_base_strategy()
        .ok()
        .map(|dirs| dirs.cache_dir().join("rv"))
}

#[inline]
fn get_packages_timeout() -> u64 {
    if let Ok(v) = std::env::var(PACKAGE_TIMEOUT_ENV_VAR_NAME) {
        if let Ok(v2) = v.parse() {
            v2
        } else {
            // If the variable doesn't parse into a valid number, return the default one
            PACKAGE_TIMEOUT
        }
    } else {
        PACKAGE_TIMEOUT
    }
}

/// Just a basic base64 without padding
#[inline]
fn encode_repository_url(url: &str) -> String {
    STANDARD_NO_PAD.encode(url)
}

#[derive(Debug, Clone)]
pub struct PackagePaths {
    pub binary: PathBuf,
    pub source: PathBuf,
}

/// This cache doesn't load anything, it just gets paths to cached objects.
/// Cache freshness is checked when requesting a path and is only a concern for package databases.
#[derive(Debug, Clone)]
pub struct DiskCache {
    /// The cache root directory.
    /// In practice it will be the OS own cache specific directory + `rv`
    root: PathBuf,
    /// R version stored as [major, minor]
    pub r_version: [u32; 2],
    /// The current execution system info: OS, version etc.
    /// Needed for binaries
    pub system_info: SystemInfo,
    /// How long the compiled databases are considered fresh for, in seconds
    /// Defaults to 3600s (1 hour)
    packages_timeout: u64,
    // TODO: check if it's worth keeping a hashmap of repo_url -> encoded
    // TODO: or if the overhead is the same as base64 directly
}

impl DiskCache {
    /// Instantiate our cache abstraction.
    pub fn new(r_version: &Version, system_info: SystemInfo) -> Result<Self> {
        let root =
            get_user_cache_dir().ok_or_else(|| anyhow!("Could not get user cache directory"))?;
        fs::create_dir_all(&root)?;
        cachedir::ensure_tag(&root).context("Failed to create CACHEDIR.TAG")?;

        Ok(Self {
            root,
            system_info,
            r_version: r_version.major_minor(),
            packages_timeout: get_packages_timeout(),
        })
    }

    /// PACKAGES databases as well as binary packages are dependent on the OS and R version
    fn get_repo_root_binary_dir(&self, repo_url: &str) -> PathBuf {
        let encoded = encode_repository_url(repo_url);
        self.root
            .join(&encoded)
            .join(get_os_path(&self.system_info, self.r_version))
    }

    /// A database contains both source and binary PACKAGE data
    /// Therefore the path to the db file is dependent on the system info and R version
    /// In practice it looks like: `CACHE_DIR/rv/{os}/{distrib?}/{arch?}/r_maj.r_min/packages.bin`
    fn get_package_db_path(&self, repo_url: &str) -> PathBuf {
        let base_path = self.get_repo_root_binary_dir(repo_url);
        base_path.join(PACKAGE_DB_FILENAME)
    }

    /// Gets the folder where a binary package would be located.
    /// The folder may or may not exist depending on whether it's in the cache
    pub fn get_binary_package_path(&self, repo_url: &str, name: &str, version: &str) -> PathBuf {
        self.get_repo_root_binary_dir(repo_url)
            .join(name)
            .join(version)
    }

    /// Gets the folder where a source tarball would be located
    /// The folder may or may not exist depending on whether it's in the cache
    pub fn get_source_package_path(&self, repo_url: &str, name: &str, version: &str) -> PathBuf {
        let encoded = encode_repository_url(repo_url);
        self.root.join(encoded).join("src").join(name).join(version)
    }

    pub fn get_package_paths(&self, repo_url: &str, name: &str, version: &str) -> PackagePaths {
        PackagePaths {
            source: self.get_source_package_path(repo_url, name, version),
            binary: self.get_binary_package_path(repo_url, name, version),
        }
    }
}

impl Cache for DiskCache {
    fn get_package_db_entry(&self, repo_url: &str) -> CacheEntry {
        let path = self.get_package_db_path(repo_url);
        if path.exists() {
            let created = path
                .metadata()
                .expect("to work")
                .created()
                .expect("to have a creation time");
            let now = SystemTime::now();

            return if now.duration_since(created).unwrap_or_default().as_secs()
                > self.packages_timeout
            {
                CacheEntry::Expired(path)
            } else {
                CacheEntry::Existing(path)
            };
        }

        CacheEntry::NotFound(path)
    }

    fn get_package_installation_status(
        &self,
        repo_url: &str,
        name: &str,
        version: &str,
    ) -> InstallationStatus {
        let source_present = self
            .get_source_package_path(repo_url, name, version)
            .join(name)
            .is_dir();
        let binary_present = self
            .get_binary_package_path(repo_url, name, version)
            .join(name)
            .is_dir();

        match (source_present, binary_present) {
            (true, true) => InstallationStatus::Both,
            (true, false) => InstallationStatus::Source,
            (false, true) => InstallationStatus::Binary,
            (false, false) => InstallationStatus::Absent,
        }
    }

    fn get_git_clone_path(&self, git_url: &str) -> PathBuf {
        let encoded = encode_repository_url(git_url);
        self.root.join("git").join(encoded)
    }
}
