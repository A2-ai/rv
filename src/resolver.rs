use std::borrow::Cow;
use std::collections::{HashMap, HashSet, VecDeque};
use std::path::Path;
use std::{fmt, fs};

use crate::cache::InstallationStatus;
use crate::config::DependencyKind;
use crate::consts::DESCRIPTION_FILENAME;
use crate::git::GitReference;
use crate::lockfile::{LockedPackage, Lockfile, Source};
use crate::package::{parse_description_file, InstallationDependencies, Package, PackageType};
use crate::repository::RepositoryDatabase;
use crate::version::{Version, VersionRequirement};
use crate::{Cache, GitOperations};

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct ResolvedDependency<'d> {
    pub(crate) name: Cow<'d, str>,
    pub(crate) version: Cow<'d, str>,
    pub(crate) source: Source,
    pub(crate) dependencies: Vec<Cow<'d, str>>,
    pub(crate) suggests: Vec<Cow<'d, str>>,
    pub(crate) force_source: bool,
    pub(crate) install_suggests: bool,
    pub(crate) kind: PackageType,
    pub(crate) installation_status: InstallationStatus,
    pub(crate) path: Option<Cow<'d, str>>,
    pub(crate) found_in_lockfile: bool,
}

impl<'d> ResolvedDependency<'d> {
    pub fn is_installed(&self) -> bool {
        match self.kind {
            PackageType::Source => self.installation_status.source_available(),
            PackageType::Binary => self.installation_status.binary_available(),
        }
    }

    pub fn from_locked_package(package: &'d LockedPackage, cache: &'d impl Cache) -> Self {
        Self {
            name: Cow::Borrowed(&package.name),
            version: Cow::Borrowed(&package.version),
            source: package.source.clone(),
            dependencies: package
                .dependencies
                .iter()
                .map(|d| Cow::Borrowed(d.as_str()))
                .collect(),
            suggests: package
                .suggests
                .iter()
                .map(|s| Cow::Borrowed(s.as_str()))
                .collect(),
            // TODO: what should we do here?
            kind: if package.force_source {
                PackageType::Source
            } else {
                PackageType::Binary
            },
            force_source: package.force_source,
            install_suggests: package.install_suggests(),
            installation_status: cache.get_package_installation_status(
                package.source.repository_url(),
                &package.name,
                &package.version,
            ),
            path: package.path.as_ref().map(|x| Cow::Borrowed(x.as_str())),
            found_in_lockfile: true,
        }
    }

    // TODO: 2 bool not great but maybe ok if it's only used in one place
    pub fn from_package_repository(
        package: &'d Package,
        repo_url: &str,
        package_type: PackageType,
        install_suggestions: bool,
        force_source: bool,
        cache: &'d impl Cache,
    ) -> (Self, InstallationDependencies<'d>) {
        let deps = package.dependencies_to_install(install_suggestions);

        let res = ResolvedDependency {
            name: Cow::Borrowed(&package.name),
            version: Cow::Borrowed(&package.version.original),
            source: Source::Repository {
                repository: repo_url.to_string(),
            },
            dependencies: deps
                .direct
                .iter()
                .map(|d| Cow::Borrowed(d.name()))
                .collect(),
            suggests: deps
                .suggests
                .iter()
                .map(|d| Cow::Borrowed(d.name()))
                .collect(),
            kind: package_type,
            force_source,
            install_suggests: install_suggestions,
            installation_status: cache.get_package_installation_status(
                repo_url,
                &package.name,
                &package.version.original,
            ),
            path: package.path.as_ref().map(|x| Cow::Borrowed(x.as_str())),
            found_in_lockfile: false,
        };

        (res, deps)
    }

    pub fn from_git_package<'p>(
        package: &'p Package,
        repo_url: &str,
        sha: String,
        install_suggestions: bool,
        cache: &'d impl Cache,
    ) -> (Self, InstallationDependencies<'p>) {
        let deps = package.dependencies_to_install(install_suggestions);

        let res = Self {
            dependencies: deps
                .direct
                .iter()
                .map(|d| Cow::Owned(d.name().to_string()))
                .collect(),
            suggests: deps
                .suggests
                .iter()
                .map(|s| Cow::Owned(s.name().to_string()))
                .collect(),
            kind: PackageType::Source,
            force_source: true,
            install_suggests: install_suggestions,
            installation_status: cache.get_git_installation_status(repo_url, &sha),
            path: None,
            found_in_lockfile: false,
            name: Cow::Owned(package.name.clone()),
            version: Cow::Owned(package.version.original.clone()),
            source: Source::Git {
                git: repo_url.to_string(),
                commit: Some(sha),
                tag: None,
                branch: None,
                directory: None,
            },
        };

        (res, deps)
    }
}

impl<'a> fmt::Display for ResolvedDependency<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}={} ({}, type={}, from_lockfile={}, path='{}')",
            self.name,
            self.version,
            self.source,
            self.kind,
            self.found_in_lockfile,
            self.path.as_deref().unwrap_or("")
        )
    }
}

#[derive(Debug, PartialEq, Clone)]
enum UnresolvedDependencyKind<'d> {
    /// The user provided a dependency that doesn't exist
    Direct,
    /// A package has a dependency not found. It could be nested several times,
    /// we only show the immediate parent which could be an indirect dep as well.
    Indirect(Cow<'d, str>),
}

#[derive(Debug, PartialEq, Clone)]
pub struct UnresolvedDependency<'d> {
    name: Cow<'d, str>,
    // source: Option<&'d Source>,
    // error: Option<String>,
    version_requirement: Option<VersionRequirement>,
    origins: Vec<UnresolvedDependencyKind<'d>>,
}

impl<'a> fmt::Display for UnresolvedDependency<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut origins = Vec::with_capacity(self.origins.len());
        for origin in &self.origins {
            let v = match origin {
                UnresolvedDependencyKind::Direct => "user provided".to_string(),
                UnresolvedDependencyKind::Indirect(parent) => format!("dependency of `{parent}`"),
            };
            origins.push(v);
        }

        write!(
            f,
            "{}{} {}",
            self.name,
            if let Some(l) = &self.version_requirement {
                format!(" {l} ")
            } else {
                String::new()
            },
            format!("from: {}", origins.join(", "))
        )
    }
}

fn read_local_description_file(
    folder: impl AsRef<Path>,
) -> Result<Package, Box<dyn std::error::Error>> {
    let folder = folder.as_ref();
    let description_path = folder.join(DESCRIPTION_FILENAME);

    match fs::read_to_string(&description_path) {
        Ok(content) => {
            if let Some(package) = parse_description_file(&content) {
                Ok(package)
            } else {
                Err(format!("Invalid DESCRIPTION file at {}", description_path.display()).into())
            }
        }
        Err(e) => Err(format!(
            "Could not read destination file at {} {e}",
            description_path.display()
        )
        .into()),
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ResolutionNeeded {
    /// Nothing to do
    None,
    /// We only need to remove those deps from the lockfile + project library
    RemoveOnly(Vec<String>),
    /// We will need to look up the package databases so do a full lookup, preferring
    /// versions already in the lockfile
    Full,
}

#[derive(Debug, PartialEq)]
pub struct Resolver<'d> {
    /// The repositories are stored in the order defined in the config
    /// The last should get priority over previous repositories
    /// If the bool is `true`, this means this repository should only look at sources
    repositories: &'d [(RepositoryDatabase, bool)],
    r_version: &'d Version,
    /// If we have a lockfile for the resolver, we will skip looking at the database for any package
    /// listed in it
    lockfile: Option<&'d Lockfile>,
}

impl<'d> Resolver<'d> {
    pub fn new(
        repositories: &'d [(RepositoryDatabase, bool)],
        r_version: &'d Version,
        lockfile: Option<&'d Lockfile>,
    ) -> Self {
        Self {
            repositories,
            r_version,
            lockfile,
        }
    }

    /// Tries to find all dependencies from the repos, as well as their install status
    pub fn resolve(
        &self,
        dependencies: &'d [DependencyKind],
        cache: &'d impl Cache,
        git_ops: &'d impl GitOperations,
    ) -> (Vec<ResolvedDependency<'d>>, Vec<UnresolvedDependency<'d>>) {
        let mut resolved = Vec::new();
        // We might have the same unresolved dep multiple times.
        let mut unresolved = HashMap::<_, UnresolvedDependency>::new();
        let mut found = HashSet::with_capacity(dependencies.len() * 10);

        let mut queue: VecDeque<_> = dependencies
            .iter()
            .map(|d| {
                (
                    Cow::Borrowed(d.name()),
                    d.as_lockfile_source(),
                    // required version
                    None,
                    d.install_suggestions(),
                    d.force_source(),
                    None,
                )
            })
            .collect();

        while let Some((
            name,
            pkg_source,
            version_requirement,
            install_suggestions,
            pkg_force_source,
            parent,
        )) = queue.pop_front()
        {
            // If we have already found that dependency, skip it
            // TODO: maybe different version req? we can cross that bridge later
            if found.contains(name.as_ref()) {
                continue;
            }

            // Look at lockfile before looking up any repositories
            if let Some(lockfile) = self.lockfile {
                // If we found the package in the lockfile, consider it found and do not look up
                // the repo at all
                if let Some(package) = lockfile.get_package(
                    name.as_ref(),
                    pkg_force_source,
                    install_suggestions,
                    pkg_source.as_ref(),
                ) {
                    found.insert(name.to_string());
                    resolved.push(ResolvedDependency::from_locked_package(package, cache));

                    for d in package.dependencies.iter().chain(&package.suggests) {
                        if !found.contains(d.as_str()) {
                            queue.push_back((
                                Cow::Borrowed(d.as_str()),
                                None,
                                None,
                                false,
                                false,
                                Some(name.clone()),
                            ));
                        }
                    }

                    continue;
                }
            }

            // For git and local sources, we could have errors (eg impossible to access repo, folder
            // not found) but we want to try to resolve everything rather than returning early
            let mut error = None;
            match pkg_source {
                None | Some(Source::Repository { .. }) => {
                    let repository = pkg_source.and_then(|c| c.r_repository());

                    for (repo, repo_source_only) in self.repositories {
                        if let Some(ref r) = repository {
                            if repo.name != *r {
                                continue;
                            }
                        }

                        if let Some((package, package_type)) = repo.find_package(
                            name.as_ref(),
                            version_requirement.as_ref(),
                            self.r_version,
                            pkg_force_source || *repo_source_only,
                        ) {
                            found.insert(name.to_string());
                            let (resolved_dep, deps) = ResolvedDependency::from_package_repository(
                                package,
                                &repo.url,
                                package_type,
                                install_suggestions,
                                pkg_force_source || *repo_source_only,
                                cache,
                            );

                            // deps.suggests will be empty if we don't have install_suggests=True
                            for d in deps.direct.into_iter().chain(deps.suggests) {
                                if !found.contains(d.name()) {
                                    queue.push_back((
                                        Cow::Borrowed(d.name()),
                                        None,
                                        d.version_requirement().map(|v| v.clone()),
                                        false,
                                        false,
                                        Some(name.clone()),
                                    ));
                                }
                            }
                            resolved.push(resolved_dep);
                            break;
                        }
                    }
                }
                // For local and git deps we assume they are source and will need to read the
                // DESCRIPTION file to get their dependencies
                Some(Source::Local { path }) => match read_local_description_file(path) {
                    Ok(package) => {
                        let deps = package.dependencies_to_install(install_suggestions);
                        found.insert(package.name);
                        todo!("handle local deps");
                    }
                    Err(e) => {
                        error = Some(format!("{e}"));
                    }
                },
                Some(Source::Git {
                    ref git,
                    ref tag,
                    ref branch,
                    ref commit,
                    directory,
                }) => {
                    let git_ref = if let Some(c) = commit {
                        GitReference::Commit(c)
                    } else if let Some(b) = branch {
                        GitReference::Branch(b)
                    } else if let Some(t) = tag {
                        GitReference::Tag(t)
                    } else {
                        unreachable!("Got an empty git reference")
                    };

                    let clone_path = cache.get_git_clone_path(git);
                    match git_ops.clone_and_checkout(git, git_ref.clone(), &clone_path) {
                        Ok(sha) => match read_local_description_file(clone_path) {
                            Ok(package) => {
                                let (resolved_dep, deps) = ResolvedDependency::from_git_package(
                                    &package,
                                    &git,
                                    sha,
                                    install_suggestions,
                                    cache,
                                );

                                // deps.suggests will be empty if we don't have install_suggests=True
                                for d in deps.direct.into_iter().chain(deps.suggests) {
                                    if !found.contains(d.name()) {
                                        queue.push_back((
                                            Cow::Owned(d.name().to_string()),
                                            None,
                                            d.version_requirement().map(|v| v.clone()),
                                            false,
                                            false,
                                            Some(name.clone()),
                                        ));
                                    }
                                }
                                resolved.push(resolved_dep);
                                continue;
                            }
                            Err(e) => {
                                error = Some(format!("{e}"));
                            }
                        },
                        Err(e) => {
                            error = Some(format!(
                                "Could not clone repository {git} (ref: {:?}) {e}",
                                git_ref,
                            ))
                        }
                    }
                }
            }

            // We don't look at repo for git and local packages
            if !found.contains(name.as_ref()) {
                let ud_kind = if let Some(p) = parent {
                    UnresolvedDependencyKind::Indirect(p)
                } else {
                    UnresolvedDependencyKind::Direct
                };
                if let Some(ud) = unresolved.get_mut(&name) {
                    ud.origins.push(ud_kind);
                } else {
                    unresolved.insert(
                        name.clone(),
                        UnresolvedDependency {
                            name,
                            version_requirement,
                            origins: vec![ud_kind],
                        },
                    );
                }
            }
        }

        (resolved, unresolved.into_values().collect())
    }

    // TODO: add tests
    pub fn resolution_needed(&self, dependencies: &'d [DependencyKind]) -> ResolutionNeeded {
        // If we don't have a lockfile, we'll need to lookup everything
        if self.lockfile.is_none() {
            return ResolutionNeeded::Full;
        }
        let lockfile = self.lockfile.unwrap();
        // At this point we need to figure out 2 things:
        // 1. whether we have all the explicit deps and their deps in the lockfile
        // 2. whether we removed some dep in the config file and need to update the lockfile
        //    and just remove those from the lockfile/rv dir
        let lockfile_deps = lockfile.package_names();
        let mut deps_seen: HashSet<&str> = HashSet::new();

        for d in dependencies {
            let all_deps = lockfile.get_package_tree(
                d.name(),
                d.force_source(),
                d.install_suggestions(),
                d.as_lockfile_source().as_ref(),
            );
            // If we don't have an explicit dep in the lockfile, we'll need a full resolve
            if all_deps.is_empty() {
                return ResolutionNeeded::Full;
            }
            deps_seen.extend(all_deps.into_iter());
        }

        // Check whether we have things we need to remove or not
        let unneeded_deps = lockfile_deps
            .difference(&deps_seen)
            .map(|d| d.to_string())
            .collect::<Vec<_>>();

        if unneeded_deps.is_empty() {
            ResolutionNeeded::None
        } else {
            ResolutionNeeded::RemoveOnly(unneeded_deps)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::Config;
    use crate::repository::RepositoryDatabase;
    use crate::CacheEntry;
    use git2::Error;
    use serde::Deserialize;
    use std::path::{Path, PathBuf};
    use std::str::FromStr;

    struct FakeCache;

    impl Cache for FakeCache {
        fn get_package_db_entry(&self, _: &str) -> CacheEntry {
            CacheEntry::NotFound(PathBuf::from_str("").unwrap())
        }

        fn get_package_installation_status(&self, _: &str, _: &str, _: &str) -> InstallationStatus {
            InstallationStatus::Absent
        }

        fn get_git_installation_status(&self, _: &str, _: &str) -> InstallationStatus {
            InstallationStatus::Absent
        }

        fn get_git_clone_path(&self, _: &str) -> PathBuf {
            PathBuf::from("")
        }
    }

    struct FakeGit;

    impl GitOperations for FakeGit {
        fn clone_and_checkout(
            &self,
            url: &str,
            git_ref: GitReference<'_>,
            destination: impl AsRef<Path>,
        ) -> Result<String, Error> {
            Ok("abc".to_string())
        }
    }

    #[derive(Debug, Deserialize)]
    struct TestRepo {
        name: String,
        source: Option<String>,
        binary: Option<String>,
        force_source: bool,
    }

    #[derive(Debug, Deserialize)]
    struct TestRepositories {
        repos: Vec<TestRepo>,
    }

    fn extract_test_elements(
        path: &Path,
    ) -> (Config, Version, Vec<(RepositoryDatabase, bool)>, Lockfile) {
        let content = std::fs::read_to_string(path).unwrap();
        let parts: Vec<_> = content.splitn(3, "---").collect();
        let config = Config::from_str(parts[0]).expect("valid config");
        let r_version = config.r_version().clone();
        let repositories = if let Ok(data) = toml::from_str::<TestRepositories>(parts[1]) {
            let mut res = Vec::new();
            for r in data.repos {
                let mut repo = RepositoryDatabase::new(&r.name, &format!("http://{}", r.name));
                if let Some(p) = r.source {
                    let path = format!("src/tests/package_files/{p}.PACKAGE");
                    let text = std::fs::read_to_string(&path).unwrap();
                    repo.parse_source(&text);
                }

                if let Some(p) = r.binary {
                    let path = format!("src/tests/package_files/{p}.PACKAGE");
                    let text = std::fs::read_to_string(&path).unwrap();
                    repo.parse_binary(&text, r_version.major_minor());
                }
                res.push((repo, r.force_source));
            }
            res
        } else {
            let mut repo = RepositoryDatabase::new("inline", "");
            repo.parse_source(parts[1]);
            vec![(repo, false)]
        };
        let lockfile = if parts[2].is_empty() {
            Lockfile::new(&r_version.original)
        } else {
            Lockfile::from_str(parts[2]).expect("valid lockfile")
        };

        (config, r_version, repositories, lockfile)
    }

    #[test]
    fn resolving() {
        let paths = std::fs::read_dir("src/tests/resolution/").unwrap();
        for path in paths {
            let p = path.unwrap().path();
            let (config, r_version, repositories, lockfile) = extract_test_elements(&p);
            let resolver = Resolver::new(&repositories, &r_version, Some(&lockfile));
            let (resolved, unresolved) =
                resolver.resolve(&config.dependencies(), &FakeCache {}, &FakeGit {});
            // let new_lockfile = Lockfile::from_resolved(&r_version.major_minor(), &resolved);
            // println!("{}", new_lockfile.as_toml_string());
            let mut out = String::new();
            for d in resolved {
                out.push_str(&d.to_string());
                out.push_str("\n");
            }

            if !unresolved.is_empty() {
                out.push_str("--- unresolved --- \n");
                for d in unresolved {
                    out.push_str(&d.to_string());
                    out.push_str("\n");
                }
            }
            // Output has been compared with pkgr for the same PACKAGE file
            insta::assert_snapshot!(p.file_name().unwrap().to_string_lossy().to_string(), out);
        }
    }
}
