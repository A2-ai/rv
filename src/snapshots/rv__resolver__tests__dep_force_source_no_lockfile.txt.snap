---
source: src/resolver.rs
expression: out
---
R6=2.5.1 (repository(url: http://cran), type=source, from_lockfile=false, path='')
